File: packages/openrouter-client-py/src/openrouter_client_py/memory.py
"""supermemory - Optional memory layer for openrouter_client_py.

Store exchanges (POST /v3/documents, taskType "memory") and retrieve prior
context (POST /v4/search) scoped to ONE singular ``containerTag``
(``^[a-zA-Z0-9_:-]+$``, first write creates the tag, never cross-tag).

The API key is read from ``SUPERMEMORY_API_KEY`` in the environment only.
Every method degrades gracefully: missing key or any request failure returns
``None`` / ``[]`` so the host application never breaks.
"""

from __future__ import annotations

import asyncio
import hashlib
import json
import logging
import os
import urllib.request
from typing import Any

logger = logging.getLogger(__name__)

DEFAULT_API_BASE = "https://api.supermemory.ai"


def is_available() -> bool:
    """True when SUPERMEMORY_API_KEY is present in the environment."""
    return bool(os.environ.get("SUPERMEMORY_API_KEY"))


def _stable_id(*parts: str) -> str:
    """Stable 24-char id from parts (customId for idempotent ingests)."""
    return hashlib.sha1("|".join(parts).encode("utf-8")).hexdigest()[:24]


class SupermemoryMemory:
    """Dependency-free memory client (stdlib urllib only)."""

    def __init__(
        self,
        container_tag: str,
        api_key: str | None = None,
        api_base: str | None = None,
        timeout: float = 15.0,
    ) -> None:
        self.container_tag = container_tag
        self.api_base = api_base or os.environ.get("SUPERMEMORY_API_BASE") or DEFAULT_API_BASE
        self.api_key = api_key or os.environ.get("SUPERMEMORY_API_KEY")
        self.timeout = timeout

    def _post(self, path: str, body: dict[str, Any]) -> dict[str, Any] | None:
        if not self.api_key:
            return None
        req = urllib.request.Request(
            f"{self.api_base}{path}",
            data=json.dumps(body).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.api_key}",
            },
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=self.timeout) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def store_exchange(
        self,
        role: str,
        content: str,
        metadata: dict[str, str] | None = None,
    ) -> str | None:
        """Store one exchange with the full memory pipeline.

        Returns the stable customId, or ``None`` when unavailable/failed.
        """
        if not self.api_key or not content.strip():
            return None
        try:
            custom_id = _stable_id(self.container_tag, role, content)
            self._post(
                "/v3/documents",
                {
                    "content": f"{role}: {content}"[:100_000],
                    "containerTag": self.container_tag,
                    "taskType": "memory",
                    "customId": custom_id,
                    "metadata": {"role": role, **(metadata or {})},
                },
            )
            return custom_id
        except Exception as exc:
            logger.warning("supermemory store failed: %s", exc)
            return None

    def search_context(
        self,
        query: str,
        limit: int = 5,
        search_mode: str = "hybrid",
    ) -> list[dict[str, Any]]:
        """Search prior memories + documents.

        Returns normalized [{"id", "content", "score", "source"}] entries,
        or ``[]`` when unavailable/failed.
        """
        if not self.api_key or not query.strip():
            return []
        try:
            data = self._post(
                "/v4/search",
                {
                    "q": query,
                    "containerTag": self.container_tag,
                    "searchMode": search_mode,
                    "limit": limit,
                },
            )
            out: list[dict[str, Any]] = []
            for r in (data or {}).get("results", []) or []:
                chunk = r.get("chunk") or r.get("memory") or r.get("content")
                out.append(
                    {
                        "id": r.get("id") or r.get("memoryId"),
                        "content": chunk if isinstance(chunk, str) else json.dumps(chunk),
                        "score": float(r.get("similarity") or r.get("score") or 0.0),
                        "source": "document" if r.get("chunk") else "memory",
                    }
                )
            return out[:limit]
        except Exception as exc:
            logger.warning("supermemory search failed: %s", exc)
            return []

    @staticmethod
    def build_system_context(results: list[dict[str, Any]], max_chars: int = 2000) -> str:
        """Turn results into a compact system-context block ("" when empty)."""
        if not results:
            return ""
        lines = ["Relevant prior context from memory:"]
        used = len(lines[0])
        for r in results:
            line = f"- [{r['source']}·{r['score']:.2f}] {r['content']}"
            if used + len(line) + 1 > max_chars:
                break
            lines.append(line)
            used += len(line) + 1
        return "\n".join(lines)

    # Async wrappers (stdlib-safe, no event loop blocking).
    async def astore_exchange(
        self,
        role: str,
        content: str,
        metadata: dict[str, str] | None = None,
    ) -> str | None:
        return await asyncio.to_thread(self.store_exchange, role, content, metadata)

    async def asearch_context(
        self,
        query: str,
        limit: int = 5,
        search_mode: str = "hybrid",
    ) -> list[dict[str, Any]]:
        return await asyncio.to_thread(self.search_context, query, limit, search_mode)


__all__ = ["SupermemoryMemory", "is_available"]

File: packages/openrouter-client/src/memory.ts
/**
 * Supermemory memory layer for openrouter-client.
 *
 * Optional, dependency-free (uses global fetch). Store exchanges with
 * taskType "memory" (POST /v3/documents) and retrieve context before the
 * model answers (POST /v4/search). Scoped to ONE singular containerTag —
 * format ^[a-zA-Z0-9_:-]+$, first write creates the tag, never cross-tag.
 *
 * The API key is read from process.env.SUPERMEMORY_API_KEY only. When it is
 * absent, every method is a no-op returning null/[] so the host app never
 * crashes and behavior stays backward compatible.
 */

export interface SupermemoryMemoryOptions {
	/** API base URL (default https://api.supermemory.ai) */
	apiBase?: string;
	/** Custom key provider; defaults to process.env.SUPERMEMORY_API_KEY */
	apiKey?: () => string | undefined;
	/** Per-request timeout in ms (default 15000) */
	timeoutMs?: number;
}

export interface MemorySearchResult {
	id?: string;
	content: string;
	score: number;
	source: "memory" | "document";
}

interface MessageLike {
	role: string;
	content: string;
}

const DEFAULT_API_BASE = "https://api.supermemory.ai";

function stableHash(input: string): string {
	let h1 = 0xdeadbeef ^ 0;
	let h2 = 0x41c6ce57 ^ 0;
	for (let i = 0; i < input.length; i++) {
		const ch = input.charCodeAt(i);
		h1 = Math.imul(h1 ^ ch, 2654435761);
		h2 = Math.imul(h2 ^ ch, 1597334677);
	}
	h1 = Math.imul(h1 ^ (h1 >>> 16), 2246822507) ^ Math.imul(h2 ^ (h2 >>> 13), 3266489909);
	h2 = Math.imul(h2 ^ (h2 >>> 16), 2246822507) ^ Math.imul(h1 ^ (h1 >>> 13), 3266489909);
	return (4294967296 * (2097151 & h2) + (h1 >>> 0)).toString(36).slice(0, 24);
}

export class SupermemoryMemory {
	private containerTag: string;
	private apiBase: string;
	private apiKey: () => string | undefined;
	private timeoutMs: number;

	constructor(containerTag: string, options: SupermemoryMemoryOptions = {}) {
		this.containerTag = containerTag;
		this.apiBase = options.apiBase ?? process.env.SUPERMEMORY_API_BASE ?? DEFAULT_API_BASE;
		this.apiKey = options.apiKey ?? (() => process.env.SUPERMEMORY_API_KEY);
		this.timeoutMs = options.timeoutMs ?? 15000;
	}

	/** True when SUPERMEMORY_API_KEY is present in the environment. */
	static isAvailable(): boolean {
		return Boolean(process.env.SUPERMEMORY_API_KEY);
	}

	/**
	 * Store one exchange (taskType "memory", full extraction pipeline).
	 * Returns the created document id, or null when the key is missing or the
	 * request fails (errors are swallowed — memory must never break chat).
	 */
	async storeExchange(
		role: "system" | "user" | "assistant" | "tool",
		content: string,
		metadata: Record<string, string> = {},
	): Promise<string | null> {
		const key = this.apiKey();
		if (!key || !content.trim()) return null;
		try {
			const customId = stableHash(`${this.containerTag}|${role}|${content}`);
			const body = {
				content: `${role}: ${content}`.slice(0, 100_000),
				containerTag: this.containerTag,
				taskType: "memory",
				customId,
				metadata: { role, ...metadata },
			};
			await this._post("/v3/documents", body, key);
			return customId;
		} catch {
			return null;
		}
	}

	/**
	 * Search prior context for a query (memories + documents).
	 * Returns normalized results, or [] when unavailable/failed.
	 */
	async searchContext(
		query: string,
		options: { limit?: number; searchMode?: "memories" | "documents" | "hybrid" } = {},
	): Promise<MemorySearchResult[]> {
		const key = this.apiKey();
		if (!key || !query.trim()) return [];
		try {
			const data = await this._post(
				"/v4/search",
				{
					q: query,
					containerTag: this.containerTag,
					searchMode: options.searchMode ?? "hybrid",
					limit: options.limit ?? 5,
				},
				key,
			);
			const results: MemorySearchResult[] = (data.results ?? []).map((r: any) => {
				const chunk = r.chunk ?? r.memory ?? r.content;
				return {
					id: r.id ?? r.memoryId ?? undefined,
					content: typeof chunk === "string" ? chunk : JSON.stringify(chunk),
					score: Number(r.similarity ?? r.score ?? 0),
					source: r.chunk ? "document" : "memory",
				};
			});
			return results.slice(0, options.limit ?? 5);
		} catch {
			return [];
		}
	}

	/** Turn search results into a compact system-context block. */
	static buildSystemContext(results: MemorySearchResult[], maxChars = 2000): string {
		if (results.length === 0) return "";
		let out = "Relevant prior context from memory:\n";
		let used = out.length;
		for (const r of results) {
			const line = `- [${r.source}·${r.score.toFixed(2)}] ${r.content}\n`;
			if (used + line.length > maxChars) break;
			out += line;
			used += line.length;
		}
		return out;
	}

	/**
	 * Inject retrieved context into a message list as (or merged into) the
	 * system message. Returns a NEW array; the input is not mutated.
	 */
	injectContext(
		messages: MessageLike[],
		query: string,
		options: { limit?: number; searchMode?: "memories" | "documents" | "hybrid" } = {},
	): Promise<MessageLike[]> {
		return (async () => {
			const results = await this.searchContext(query, options);
			const block = SupermemoryMemory.buildSystemContext(results);
			if (!block) return messages;
			const copy: MessageLike[] = [...messages];
			const sysIdx = copy.findIndex((m) => m.role === "system");
			if (sysIdx === -1) {
				copy.unshift({ role: "system", content: block });
			} else {
				copy[sysIdx] = { ...copy[sysIdx], content: `${copy[sysIdx].content}\n\n${block}` };
			}
			return copy;
		})();
	}

	private async _post(path: string, body: unknown, key: string): Promise<any> {
		const controller = new AbortController();
		const timer = setTimeout(() => controller.abort(), this.timeoutMs);
		try {
			const res = await fetch(`${this.apiBase}${path}`, {
				method: "POST",
				headers: {
					"Content-Type": "application/json",
					Authorization: `Bearer ${key}`,
				},
				body: JSON.stringify(body),
				signal: controller.signal,
			});
			if (!res.ok) {
				throw new Error(`supermemory ${path} -> HTTP ${res.status}`);
			}
			return await res.json();
		} finally {
			clearTimeout(timer);
		}
	}
}

export default SupermemoryMemory;